conbench/entities/run.py

```diff
@@ -209,6 +179,117 @@ def get_baseline_id(self):
+
+    If Commit not yet known in database: fetch data about commit (and related
+    commits) from GitHub HTTP API if possible. Exceptions during this process
+    are logged and otherwise swallowed.
+
+    Return Commit.id (DB primary key) of existing Commit entity or of newly
+    created one. Expect database collision upon insert (in this case the ID for
+    the existing commit entity is returned).
+
+    Has slightly ~unpredictable run duration as of interaction with GitHub HTTP
+    API.
+    """
+
+    def _guts(sha, repository, pr_number, branch) -> Commit:
+        """
+        Return a Commit object or raise `sqlalchemy.exc.IntegrityError`.
+        """
+        # Try to see if commit is already database. This is an optimization, to
+        # not needlessly interact with the GitHub HTTP API in case the commit
+        # is already in the database. first(): "Return the first result of this
+        # Query or None if the result doesn’t contain any row.""
+        commit: Optional[Commit] = Commit.first(sha=sha, repository=repository)
+
+        if commit is not None:
+            return commit
+
+        # Try to fetch metadata for commit via GitHub HTTP API. Fall back
+        # gracefully if that does not work.
+
+        gh_commit_dict = None
+        try:
+            # get_github_commit() may raise all those exceptions that can
+            # happen during an HTTP request cycle.
+            gh_commit_dict = get_github_commit(
+                repository=repository, pr_number=pr_number, branch=branch, sha=sha
+            )
+        except Exception as exc:
+            log.info(
+                "treat as unknown commit: error during get_github_commit(): %s", exc
+            )
+
+        if gh_commit_dict:
+            # We got data from GitHub. Insert into database.
+            commit = Commit.create_github_context(sha, repository, gh_commit_dict)
+
+            # The commit is known to GitHub. Fetch more data from GitHub.
+            # Gracefully degradate if that does not work.
+            try:
+                backfill_default_branch_commits(repository, commit)
+            except Exception as exc:
+                # Any error during this backfilling operation should not fail
+                # the HTTP request processing (we're right now in the middle of
+                # processing an HTTP request with new benchmark run data).
+                log.info(
+                    "Could not backfill default branch commits. Ignoring error "
+                    "during backfill_default_branch_commits():  %s",
+                    exc,
+                )
+
+        # If at least a commit hash and/or repository are provided, insert
+        # that. Note(JP): wouldn't it be better if we hat the constraint that a
+        # commit entity in the database has _at least_ repository specifier and
+        # commit hash? What is the value of a commit object that only refers to
+        # a repository (and not to a commit hash)?
```

Comment:
@jonkeane @austin3dickey would love to have quick opinion, thanks!
let's put findings here: https://github.com/conbench/conbench/issues/817

Reply:
I'm fine with that and can't think of bad things that would happen if it didn't. The original motivation for the looseness here was to also allow us to store results on other softwares (some where we might not have a specific hash that we know we're running). We haven't done that so far, and I suspect if (when?) we do, we'll want to create a separate table for "external benchmark" that has different constraints so we can deal with that there (or we can loosen this back if we get there.

Reply:
After #817 I have now removed the comment and updated the code to be clear about the fact that here, both `sha` and `repository` are non-None.